### PR TITLE
Mpc integration

### PR DESCRIPTION
**Pull request by ML1907**
Default NMPC controller for SVEA platform. 
Release 0 features: 
- Non linear vehicle dynamics implemented.
- steering rate and acceleration as MPC inputs.
- produces an optimal path plan and optimal control inputs.
- does NOT handle neither static nor dynamic obstacles.
- adaptive prediction horizon
- customizable running frequency

To DOs:
- Code refactoring
- Integration into SVEA main repository
- further testing with Mocap

### CODE REVIEW

File: .gitignore
Comment (kaarmu) [inline .gitignore:1]: do not include

File: .gitmodules
Comment (kaarmu) [inline .gitmodules:1]: do not include

File: src/svea_sensors/params/robot_localization/local_ekf.yaml
Comment (kaarmu) [inline src/svea_sensors/params/robot_localization/local_ekf.yaml:1]: do not include

File: src/svea_sensors/scripts/set_datum_node.py
Comment (kaarmu) [inline src/svea_sensors/scripts/set_datum_node.py:1]: do not include

File: src/svea_vision
Comment (kaarmu) [inline src/svea_vision:1]: do not include

File: util/config.sh
Comment (kaarmu) [inline util/config.sh:1]: do not include

File: util/run
Comment (kaarmu) [inline util/run:1]: do not include

File: util/start
Comment (kaarmu) [inline util/start:1]: do not include

File: src/svea_core/src/svea/simulators/sim_SVEA.py
Comment (kaarmu) [inline src/svea_core/src/svea/simulators/sim_SVEA.py:1]: do not include; if important, it should be a different PR

File: src/svea_core/src/svea/simulators/sim_lidar.py
Comment (kaarmu) [inline src/svea_core/src/svea/simulators/sim_lidar.py:1]: do not include; if important, it should be a different PR

File: src/svea_core/src/svea/svea_managers/svea_archetypes.py
Comment (kaarmu) [inline src/svea_core/src/svea/svea_managers/svea_archetypes.py:1]: Is this API change supported by PurePursuitController?

File: src/svea_mocap
Comment (kaarmu) [inline src/svea_mocap:1]: do not include

File: src/svea_navigation/maps/itrl_sidewalk.pgm
Comment (kaarmu) [inline src/svea_navigation/maps/itrl_sidewalk.pgm:1]: do not include

File: src/svea_navigation/maps/itrl_sidewalk.yaml
Comment (kaarmu) [inline src/svea_navigation/maps/itrl_sidewalk.yaml:1]: do not include

File: src/svea_navigation/maps/obstacles.yaml
Comment (kaarmu) [inline src/svea_navigation/maps/obstacles.yaml:1]: do not include

File: src/svea_navigation/svea_navigation_mpc/params/mpc_params.yaml
Comment (kaarmu) [inline src/svea_navigation/svea_navigation_mpc/params/mpc_params.yaml:1]: move to `svea_core/params`

File: src/svea_navigation/svea_navigation_mpc/scripts/mpc_casadi.py
Comment (kaarmu) [inline src/svea_navigation/svea_navigation_mpc/scripts/mpc_casadi.py:1]: Move to `svea_core/src/svea/controllers/mpc.py`. Will look at implementation in the next review.

Comment (kaarmu) [reply src/svea_navigation/svea_navigation_mpc/scripts/mpc_casadi.py:1]: Also, do we need the `/set_goal_position` service or is `/mpc_target` enough?

**Review by kaarmu — CHANGES_REQUESTED** [review]
Nice work Marco!! ❇️

Generally, you can remove changes outside of `src/svea_core` and `src/svea_navigation/svea_navigation_mpc`. For example, I would just checkout complete directories like: `git checkout -- src/svea_sensors`. There is one important comment in `svea_core` and, in `svea_navigation_mpc` I commented where I think the code should go.

**Comment by ML1907** [comment]
Thanks for the feedback Kaj!!!! 🔥
________________________________
Da: Kaj Munhoz Arfvidsson ***@***.***>
Inviato: Tuesday, November 12, 2024 9:37:29 PM
A: KTH-SML/svea ***@***.***>
Cc: Marco Lodetti ***@***.***>; Author ***@***.***>
Oggetto: Re: [KTH-SML/svea] Mpc integration (PR #42)


@kaarmu requested changes on this pull request.

Nice work Marco!! ❇️

Generally, you can remove changes outside of src/svea_core and src/svea_navigation/svea_navigation_mpc. For example, I would just checkout complete directories like: git checkout -- src/svea_sensors. There is one important comment in svea_core and, in svea_navigation_mpc I commented where I think the code should go.

________________________________

On .gitignore<https://github.com/KTH-SML/svea/pull/42#discussion_r1838712935>:

do not include

________________________________

On .gitmodules<https://github.com/KTH-SML/svea/pull/42#discussion_r1838713233>:

do not include

________________________________

On docker/Dockerfile<https://github.com/KTH-SML/svea/pull/42#discussion_r1838713595>:

do not include

________________________________

On requirements.txt<https://github.com/KTH-SML/svea/pull/42#discussion_r1838714444>:

remove comments, whitespace and utm dependency

________________________________

On src/motion_capture_system<https://github.com/KTH-SML/svea/pull/42#discussion_r1838714673>:

do not include

________________________________

On src/svea_sensors/launch/localize.launch<https://github.com/KTH-SML/svea/pull/42#discussion_r1838715706>:

do not include

________________________________

On src/svea_sensors/launch/sensors_launch/lidar.launch<https://github.com/KTH-SML/svea/pull/42#discussion_r1838715845>:

do not include

________________________________

On src/svea_sensors/launch/sensors_launch/navsat.launch<https://github.com/KTH-SML/svea/pull/42#discussion_r1838715959>:

do not include

________________________________

On src/svea_sensors/launch/transforms.launch<https://github.com/KTH-SML/svea/pull/42#discussion_r1838716136>:

do not include

________________________________

On src/svea_sensors/package.xml<https://github.com/KTH-SML/svea/pull/42#discussion_r1838716232>:

do not include

________________________________

On src/svea_sensors/params/robot_localization/local_ekf.yaml<https://github.com/KTH-SML/svea/pull/42#discussion_r1838716332>:

do not include

________________________________

On src/svea_sensors/scripts/set_datum_node.py<https://github.com/KTH-SML/svea/pull/42#discussion_r1838716562>:

do not include

________________________________

On src/svea_vision<https://github.com/KTH-SML/svea/pull/42#discussion_r1838716693>:

do not include

________________________________

On util/config.sh<https://github.com/KTH-SML/svea/pull/42#discussion_r1838716847>:

do not include

________________________________

On util/run<https://github.com/KTH-SML/svea/pull/42#discussion_r1838716959>:

do not include

________________________________

On util/start<https://github.com/KTH-SML/svea/pull/42#discussion_r1838717214>:

do not include

________________________________

On src/svea_core/src/svea/simulators/sim_SVEA.py<https://github.com/KTH-SML/svea/pull/42#discussion_r1838719375>:

do not include; if important, it should be a different PR

________________________________

On src/svea_core/src/svea/simulators/sim_lidar.py<https://github.com/KTH-SML/svea/pull/42#discussion_r1838720226>:

do not include; if important, it should be a different PR

________________________________

On src/svea_core/src/svea/svea_managers/svea_archetypes.py<https://github.com/KTH-SML/svea/pull/42#discussion_r1838721282>:

Is this API change supported by PurePursuitController?

________________________________

On src/svea_mocap<https://github.com/KTH-SML/svea/pull/42#discussion_r1838721562>:

do not include

________________________________

On src/svea_navigation/maps/itrl_sidewalk.pgm<https://github.com/KTH-SML/svea/pull/42#discussion_r1838721821>:

do not include

________________________________

On src/svea_navigation/maps/itrl_sidewalk.yaml<https://github.com/KTH-SML/svea/pull/42#discussion_r1838722120>:

do not include

________________________________

On src/svea_navigation/maps/obstacles.yaml<https://github.com/KTH-SML/svea/pull/42#discussion_r1838722336>:

do not include

________________________________

On src/svea_navigation/svea_navigation_mpc/params/mpc_params.yaml<https://github.com/KTH-SML/svea/pull/42#discussion_r1838726252>:

move to svea_core/params

________________________________

On src/svea_navigation/svea_navigation_mpc/scripts/mpc_casadi.py<https://github.com/KTH-SML/svea/pull/42#discussion_r1838730474>:

Move to svea_core/src/svea/controllers/mpc.py. Will look at implementation in the next review.

________________________________

On src/svea_navigation/svea_navigation_mpc/scripts/mpc_casadi.py<https://github.com/KTH-SML/svea/pull/42#discussion_r1838734373>:

Also, do we need the /set_goal_position service or is /mpc_target enough?

File: src/svea_navigation/svea_navigation_mpc/scripts/mpc_casadi.py
Comment (ML1907) [reply src/svea_navigation/svea_navigation_mpc/scripts/mpc_casadi.py:1]: /set_goal_position is the service topic used by the GoalSetterNode to set a new goal for the MPC. I created GoalSetterNode to separate the process of listening to the /mpc_target topic from the actual goal update logic in the main file. Could simplify that in the next review if needed :)

File: src/svea_core/src/svea/svea_managers/svea_archetypes.py
Comment (ML1907) [reply src/svea_core/src/svea/svea_managers/svea_archetypes.py:1]: was not! created a new class inheriting.

**Review by kaarmu — COMMENTED** [review]
I've moved, removed and refactored quite a lot of code now. Primary difference is how parameters are loaded. If a node is using mpc, we load the params file using rosparam in the launch file. Not super happy with it, but passing around a config file makes me uncomfortable. Can you give your opinion on the changes? 

Also, could you setup an example (in `svea_example`) similar to what you showed me with  clicked_point?

**Comment by ML1907** [comment]
Okay Kaj, I believe the MPC example is now well-configured, and all the code has been refactored and thoroughly commented. Let me know if you notice anything missinng before finalizing the pull request 💯

File: src/svea_core/src/svea/controllers/mpc.py [hunk 1]
```diff
@@ -7,10 +7,29 @@
 
 class MPC_casadi:
 
-    def __init__(self, vehicle_name='', config_ns='~mpc'):
+    def __init__(self, vehicle_name='', config_ns='/mpc'):
```
Comment (kaarmu) [inline src/svea_core/src/svea/controllers/mpc.py:10]: This should be '~mpc`

**Review by kaarmu — COMMENTED** [review]
Just one small fix for namespace default. Otherwise I'm ready to merge!

**Comment by ML1907** [comment]
Files updated, ready to merge as well 🫡

**Review by kaarmu — APPROVED** [review]

**Pull request merged**